### PR TITLE
Theo/pervault collateral thresholds

### PR DESCRIPTION
**Pull request by theodorebugnet**
Adds an optional custom secure collateral threshold field to vaults, which must be greater than the system one (if set).
Affects issuable_tokens calculation, collateral withdrawal limits, and token minting on redeem cancellation with reimbursement.

### CODE REVIEW

**Review by sander2 — COMMENTED** [review]
The change LGTM, but I would like to requests some tests:
- an integration test (in test_issue.rs) showing that the custom threshold is respected when requesting an actual issue
- A test for the RPC. Our rpcs aren't really tested well in general, but since we plan to rely on them more this is something that we need to improve

**Comment by inosgun** [comment]
Aha

File: crates/vault-registry/src/lib.rs [hunk 1]
```diff
@@ -861,7 +905,7 @@ impl<T: Config> Pallet<T> {
         };
 
         let is_below_threshold =
-            Pallet::<T>::is_collateral_below_secure_threshold(&new_collateral, &vault.backed_tokens()?)?;
+            Pallet::<T>::is_collateral_below_vault_secure_threshold(&new_collateral, &vault.backed_tokens()?, &vault)?;
```
Comment (sander2) [inline crates/vault-registry/src/lib.rs:908]: Would be good to remove the `is_collateral_below_secure_threshold` function since we don't use it anymore. If we leave the function around we risk using it in the future when we should instead use the new one

**Review by sander2 — CHANGES_REQUESTED** [review]
Sorry for not mentioning this before, but you will need to add a migration for the vault storage. You can find an example of a recent vault migration [here](https://github.com/interlay/interbtc/pull/571/files#diff-a51dbc80f617db9fffe259eac5cd4ce3a24cec79bf6a95c50a286d75d4ff2dedR156)

File: crates/vault-registry/src/types.rs [hunk 2]
```diff
@@ -185,6 +187,159 @@ pub mod liquidation_vault_fix {
     }
 }
 
+pub mod v4 {
```
Comment (nud3l) [inline crates/vault-registry/src/types.rs:190]: I'm not sure how exactly the migrations work. Have @sander2 or @gregdhill checked this code?

File: crates/vault-registry/src/types.rs [hunk 3]
```diff
@@ -238,6 +393,8 @@ pub struct Vault<AccountId, BlockNumber, Balance, CurrencyId: Copy> {
     /// Block height until which this Vault is banned from being used for
     /// Issue, Redeem (except during automatic liquidation) and Replace.
     pub banned_until: Option<BlockNumber>,
+    /// Custom secure collateral threshold
+    pub secure_collateral_threshold: Option<UnsignedFixedPoint>,
```
Comment (nud3l) [inline crates/vault-registry/src/types.rs:397]: Is there a particular reason this is not inserted at the end of the struct?

File: crates/vault-registry/src/types.rs [hunk 4]
```diff
@@ -453,10 +619,18 @@ impl<T: Config> RichVault<T> {
         Pallet::<T>::get_backing_collateral(&self.id())
     }
 
+    pub fn get_personal_secure_threshold(&self) -> Result<UnsignedFixedPoint<T>, DispatchError> {
```
Comment (nud3l) [inline crates/vault-registry/src/types.rs:622]:
```suggestion
    pub fn get_custom_secure_threshold(&self) -> Result<UnsignedFixedPoint<T>, DispatchError> {
```

Comment (nud3l) [reply crates/vault-registry/src/types.rs:622]: On second thought, since this is operating directly on the vault object, how about just using `set_secure_threshold` and `get_secure_threshold` since you return anyway either the global or vault specific threshold here. Naming it `get_custom..` or `get_personal..` implies it would not fetch the global one and return `None` if unset.

File: standalone/runtime/tests/test_redeem.rs [hunk 5]
```diff
@@ -2027,6 +2027,14 @@ fn get_additional_collateral(vault_id: &VaultId) {
     ));
 }
 
+fn get_a_bit_of_additional_collateral(vault_id: &VaultId) {
```
Comment (nud3l) [inline standalone/runtime/tests/test_redeem.rs:2030]: Minor nit: I'd move this small helper directly into the test. That'll help a bit with understanding how much is added as the function isn't reused as far as I can see.

File: standalone/runtime/tests/test_vault_registry.rs [hunk 6]
```diff
@@ -227,6 +229,49 @@ mod withdraw_collateral_test {
             );
         });
     }
+
+    #[test]
+    fn integration_test_vault_registry_withdraw_collateral_respects_custom_thresholds() {
```
Comment (nud3l) [inline standalone/runtime/tests/test_vault_registry.rs:234]: Question: I think I remember that we discussed that vaults should be able to withdraw their collateral up until the global threshold. Did you guys agree to change that? I'm OK either way.

Comment (theodorebugnet) [reply standalone/runtime/tests/test_vault_registry.rs:234]: I believe the final conclusion was that the custom threshold should just be used for everything - and if a user wants to override their own threshold, they can batch a transaction to unset it, do whatever they want (including withdrawing collateral, but also e.g. self-issuing), then setting their threshold back.
This simplifies things because we don't need for example two RPCs to let the UI handle collateral withdrawing.

File: crates/vault-registry/src/types.rs (hunk 3 above)
Comment (theodorebugnet) [reply crates/vault-registry/src/types.rs:397]: Not really, but is there a particular reason it should be? Happy to move it, I'm not aware of it making a difference

File: crates/vault-registry/src/types.rs (hunk 2 above)
Comment (theodorebugnet) [reply crates/vault-registry/src/types.rs:190]: Nope, maybe then hold off on merging until one of them has had a chance to review next week.

File: crates/vault-registry/src/lib.rs [hunk 7]
```diff
@@ -131,6 +131,10 @@ pub mod pallet {
             log::info!("Off-chain worker started on block {:?}", n);
             Self::_offchain_worker();
         }
+
+        fn on_runtime_upgrade() -> frame_support::weights::Weight {
```
Comment (sander2) [inline crates/vault-registry/src/lib.rs:135]: To actually perform the migration, you need to bump the spec versions. I'm not sure what other changes this will be batched with, so I'm not sure if that should be done here or in a separate pr. @gregdhill ?

Comment (gregdhill) [reply crates/vault-registry/src/lib.rs:135]: Let's do that in a separate PR :)

**Review by gregdhill — APPROVED** [review]
